src/System.Private.CoreLib/shared/System/String.Manipulation.cs

```diff
@@ -341,6 +341,55 @@ public static string Concat(string str0, string str1, string str2, string str3)
             return result;
         }
 
+        // TODO: Expose publicly: https://github.com/dotnet/corefx/issues/34330.
+        internal static string Concat(ReadOnlySpan<char> str0, ReadOnlySpan<char> str1)
+        {
+            string result = FastAllocateString(str0.Length + str1.Length);
+            Span<char> resultSpan = new Span<char>(ref result.GetRawStringData(), result.Length);
+
+            str0.CopyTo(resultSpan);
+            str1.CopyTo(resultSpan.Slice(str0.Length));
+
+            return result;
+        }
+
+        // TODO: Expose publicly: https://github.com/dotnet/corefx/issues/34330.
+        internal static string Concat(ReadOnlySpan<char> str0, ReadOnlySpan<char> str1, ReadOnlySpan<char> str2)
+        {
+            string result = FastAllocateString(str0.Length + str1.Length + str2.Length);
+            Span<char> resultSpan = new Span<char>(ref result.GetRawStringData(), result.Length);
+
+            str0.CopyTo(resultSpan);
+            resultSpan = resultSpan.Slice(str0.Length);
+
+            str1.CopyTo(resultSpan);
+            resultSpan = resultSpan.Slice(str1.Length);
+
+            str2.CopyTo(resultSpan);
+
+            return result;
+        }
+
+        // TODO: Expose publicly: https://github.com/dotnet/corefx/issues/34330.
+        internal static string Concat(ReadOnlySpan<char> str0, ReadOnlySpan<char> str1, ReadOnlySpan<char> str2, ReadOnlySpan<char> str3)
+        {
+            string result = FastAllocateString(str0.Length + str1.Length + str2.Length + str3.Length);
+            Span<char> resultSpan = new Span<char>(ref result.GetRawStringData(), result.Length);
+
+            str0.CopyTo(resultSpan);
```

Comment:
If the  length overflow, this will throw ArgumentException with `Destination is too short.` or `message that is not very helpful  to give you a hint about the actual problem.

Maybe compute the buffer length inside `checked`? It will produce `OverflowException` that hints better at the actual problem, and allow us to skip the bounds checks in the rest of the method if we wanted to make this as fast as possible?

Reply:
Ok.